flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/ExactlyOnceKafkaWriter.java

```diff
@@ -293,22 +309,38 @@ private TransactionAbortStrategyContextImpl getTransactionAbortStrategyContext(
                     producerPool.recycle(producer);
                     return epoch;
                 };
-        Set<String> ongoingTransactionIds =
+        Set<String> precommittedTransactionalIds =
                 recoveredStates.stream()
                         .flatMap(
                                 s ->
-                                        s.getOngoingTransactions().stream()
+                                        s.getPrecommittedTransactionalIds().stream()
                                                 .map(CheckpointTransaction::getTransactionalId))
                         .collect(Collectors.toSet());
+        Set<Integer> oldSubtaskIds =
+                recoveredStates.stream()
+                        .map(KafkaWriterState::getSubtaskId)
+                        .collect(Collectors.toSet());
+        Integer oldParallelism =
+                recoveredStates.stream()
+                        .map(KafkaWriterState::getParallelism)
+                        .findFirst()
+                        .orElse(null);
+        // Special case: recovery without checkpoint -> first subtask owns everything
+        if (recoveredStates.isEmpty() && kafkaSinkContext.getParallelInstanceId() == 0) {
+            oldSubtaskIds = Set.of(0);
+            oldParallelism = 1;
+        }
```

Comment:
I think the block is easier to understand with a dedicated method that returns immediately on the special condition. It also avoids reassigning the variables